### PR TITLE
[SNOW-1489960, SNOW-1491199] Snowpark python API client-side changes for IR encoding to protobuf

### PR DESCRIPTION
**Pull request by sfc-gh-lspiegelberg**
Main commit for [SNOW-1489960](https://snowflakecomputing.atlassian.net/browse/SNOW-1489960) bringing in changes to encode Snowpark python API to protobuf. A detailed commit history can be found in https://github.com/snowflakedb/snowpark-python/tree/ls-SNOW-1491199-merge-phase0-server-side.

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

[SNOW-1489960]: https://snowflakecomputing.atlassian.net/browse/SNOW-1489960?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Co-authored-by: Arthur Zwiegincew <arthur.zwiegincew@snowflake.com>
Co-authored-by: Ovidiu Platon <ovidiu.platon@snowflake.com>
Co-authored-by: Hemit Shah <hemit.shah@snowflake.com>
Co-authored-by: Varnika Budati <varnika.budati@snowflake.com>
Co-authored-by: Eric Vandenberg <eric.vandenberg@snowflake.com>
Co-authored-by: Andong Zhan <andong.zhan@snowflake.com>

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -45,6 +47,7 @@
 DEVELOPMENT_REQUIREMENTS = [
     "pytest<8.0.0",  # check SNOW-1022240 for more details on the pin here
     "pytest-cov",
```
Comment (sfc-gh-lspiegelberg) [inline setup.py:49]: This package is needed for functools.wraps(...) from a quick search.

File: tox.ini [hunk 2]
```diff
@@ -81,6 +81,8 @@ passenv =
     SF_REGRESS_LOGS
     ; Github Actions provided environmental variables
```
Comment (sfc-gh-lspiegelberg) [inline tox.ini:82]: @sfc-gh-azhan this here changed.

Comment (sfc-gh-azhan) [reply tox.ini:82]: duplicate lines in 84 85

File: .github/workflows/precommit.yml [hunk 3]
```diff
@@ -333,6 +333,104 @@ jobs:
...
+      - name: Run doctests for AST (I/II)
+        run: python -m tox -e "py${PYTHON_VERSION}-doctest-notudf-ci"
+        env:
+          PYTHON_VERSION: ${{ matrix.python-version }}
+          cloud_provider: ${{ matrix.cloud-provider }}
+          PYTEST_ADDOPTS: --color=yes --tb=short --enable_ast
+          TOX_PARALLEL_NO_SPINNER: 1
+        shell: bash
+
+  test-ast-phase0:
```
Comment (sfc-gh-azhan) [inline .github/workflows/precommit.yml:385]: let's combine these two into one.

File: .github/workflows/precommit.yml [hunk 4]
```diff
@@ -333,6 +333,104 @@ jobs:
...
+    name: Test AST enabled (phase0) py-${{ matrix.os }}-${{ matrix.python-version }}
+    needs: build
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ ubuntu-latest ]
+        python-version: [ "3.9", "3.11" ]
```
Comment (sfc-gh-azhan) [inline .github/workflows/precommit.yml:393]: only run 3.9 in precommit.yml. Run all other options in daily_precommit.yml (This can be a separate PR).

File: scripts/tox_install_cmd.sh [hunk 5]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-set -e
+set -euxo pipefail
```
Comment (sfc-gh-azhan) [inline scripts/tox_install_cmd.sh:3]: no need to change this in this PR

File: setup.py [hunk 6]
```diff
@@ -23,6 +23,8 @@
     CONNECTOR_DEPENDENCY,
     # snowpark directly depends on typing-extension, so we should not remove it even if connector also depends on it.
     "typing-extensions>=4.1.0, <5.0.0",
+    "protobuf>=5.28",
```
Comment (sfc-gh-azhan) [inline setup.py:26]: fix it

File: setup.py [hunk 7]
```diff
@@ -45,6 +47,7 @@
 DEVELOPMENT_REQUIREMENTS = [
     "pytest<8.0.0",  # check SNOW-1022240 for more details on the pin here
     "pytest-cov",
+    "wrapt",
```
Comment (sfc-gh-azhan) [inline setup.py:50]: add comment

File: .github/workflows/precommit.yml (hunk 3 above)
Comment (sfc-gh-lspiegelberg) [reply .github/workflows/precommit.yml:385]: can you help maybe with this by filing a PR? May be faster than me experimenting with this for a few hours and then having a long ping-pong process.

File: setup.py (hunk 6 above)
Comment (sfc-gh-lspiegelberg) [reply setup.py:26]: what is the fix? I.e., what version did we settle on?

File: .github/workflows/precommit.yml (hunk 3 above)
Comment (sfc-gh-lspiegelberg) [reply .github/workflows/precommit.yml:385]: we can potentially also do this as a refactoring post-merge if you're busy today/tomorrow.

File: setup.py (hunk 6 above)
Comment (sfc-gh-azhan) [reply setup.py:26]: remove these two lines. No changes in setup.py

File: src/conftest.py [hunk 8]
```diff
@@ -12,7 +12,8 @@
 
 from snowflake.snowpark import Session
 
-logging.getLogger("snowflake.connector").setLevel(logging.ERROR)
+_logger = logging.getLogger("snowflake.connector")
```
Comment (sfc-gh-azhan) [inline src/conftest.py:15]: no change

File: .github/workflows/precommit.yml (hunk 3 above)
Comment (sfc-gh-lspiegelberg) [reply .github/workflows/precommit.yml:385]: added to followup ticket: https://snowflakecomputing.atlassian.net/browse/SNOW-1826001

File: src/snowflake/snowpark/dataframe.py [hunk 9]
```diff
@@ -520,14 +565,30 @@ class DataFrame:
...
         self._session = session
-        self._plan = self._session._analyzer.resolve(plan)
+        self._ast_id = None
+        if _emit_ast:
+            self._ast_id = _ast_stmt.var_id.bitfield1 if _ast_stmt is not None else None
+
+        if plan is not None:
+            self._plan = self._session._analyzer.resolve(plan)
+        else:
+            self._plan = None
```
Comment (sfc-gh-oplaton) [inline src/snowflake/snowpark/dataframe.py:590]: Why does the AST variant replace the original code with the `plan is not None` check?

File: src/snowflake/snowpark/dataframe.py [hunk 10]
```diff
@@ -1813,15 +2370,18 @@ def pivot(
...
         )
 
     @df_api_usage
+    @publicapi
     def unpivot(
         self,
         value_column: str,
         name_column: str,
         column_list: List[ColumnOrName],
         include_nulls: bool = False,
```
Comment (sfc-gh-oplaton) [inline src/snowflake/snowpark/dataframe.py:2383]: The existing code doesn't capture `include_nulls`, because it's a very recent addition. We need a `TODO` for this particular item. More generally, I jumped straight from `1.23` to `HEAD` when doing the latest `main` -> `phase 0` integration, so we'll need to pass over the public API surface and ensure the AST client is still capturing everything at full fidelity.

Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/dataframe.py:2383]: Yes, this has been already added to the umbrella ticket https://snowflakecomputing.atlassian.net/browse/SNOW-1826001 for post-merge consideration.

File: src/snowflake/snowpark/dataframe.py (hunk 9 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/dataframe.py:590]: good question, is this a merge gone wrong? What would be the correct code here?

Comment (sfc-gh-oplaton) [reply src/snowflake/snowpark/dataframe.py:590]: It looks like the change is intentional, and came in with one of Arthur's changes in July. Let's leave it as is.

Comment (sfc-gh-jkew) [reply src/snowflake/snowpark/dataframe.py:590]: I think pointed this out in one of the original merge PRs for a similar change in dataframe.py and the answer was that it was a bug fix

https://github.com/snowflakedb/snowpark-python/pull/2136#discussion_r1777733248

File: CHANGELOG.md [hunk 11]
```diff
@@ -19,6 +19,7 @@
     - `keyType`: keys of the map
     - `valueType`: values of the map
 - Added support for `include_nulls` argument in `DataFrame.unpivot`.
+- Added client-side thin-client mode which can be enabled with `session.ast_enabled=True` (default: `False`).
```
Comment (sfc-gh-azhan) [inline CHANGELOG.md:22]: enable ast does not enable a thin client mode right? We still send sql query to the server.

File: tox.ini (hunk 2 above)
Comment (sfc-gh-azhan) [reply tox.ini:82]: why? Such a simple change

Comment (sfc-gh-azhan) [reply tox.ini:82]: you just need to move line 84 and 85.

File: src/snowflake/snowpark/query_history.py [hunk 12]
```diff
@@ -72,3 +87,32 @@ def include_thread_id(self) -> bool:
     def include_error(self) -> bool:
         """When True, queries that have error during execution are recorded by this listener."""
         return self._include_error
+
+
+class AstListener(QueryListener):
+    def __init__(
+        self,
+        session: "snowflake.snowpark.session.Session",
+        include_failures: bool = False,
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/query_history.py:96]: add comment to talk about what `include_failures` does

File: src/snowflake/snowpark/modin/plugin/utils/numpy_to_pandas.py [hunk 13]
```diff
@@ -213,8 +213,8 @@ def map_to_bools(inputs: Any) -> Any:
     "exp": NotImplemented,
     "exp2": NotImplemented,
     "log": lambda obj, inputs: obj.apply(sp_func.ln),  # use built-in function
-    "log2": lambda obj, inputs: obj.apply(sp_func.log, base=2),
-    "log10": lambda obj, inputs: obj.apply(sp_func.log, base=10),
+    "log2": lambda obj, inputs: obj.apply(sp_func._log2),
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/utils/numpy_to_pandas.py:216]: shouldn't change

File: src/snowflake/snowpark/column.py [hunk 14]
```diff
@@ -225,108 +238,339 @@ class Column:
...
         if expr2 is not None:
-            if isinstance(expr1, str) and isinstance(expr2, str):
-                if expr2 == "*":
-                    self._expression = Star([], df_alias=expr1)
-                else:
-                    self._expression = UnresolvedAttribute(
-                        quote_name(expr2), df_alias=expr1
-                    )
-            else:
+            if not (isinstance(expr1, str) and isinstance(expr2, str)):
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/column.py:257]: we should not change existing code in this PR. We should limit the scope to add new ast logic only.

File: src/snowflake/snowpark/column.py [hunk 15]
```diff
@@ -225,108 +238,339 @@ class Column:
...
     def __getitem__(self, field: Union[str, int]) -> "Column":
         """Accesses an element of ARRAY column by ordinal position, or an element of OBJECT column by key."""
+
+        _emit_ast = self._ast is not None
+        expr = None
         if isinstance(field, str):
-            return Column(SubfieldString(self._expression, field))
+            if _emit_ast:
+                expr = proto.Expr()
+                ast = with_src_position(expr.sp_column_apply__string)
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/column.py:313]: curious we don't have a ast builder?

File: src/snowflake/snowpark/dataframe.py [hunk 16]
```diff
@@ -262,8 +301,12 @@ def _disambiguate(
     lsuffix = lsuffix or lhs._alias
     rsuffix = rsuffix or rhs._alias
     suffix_provided = lsuffix or rsuffix
-    lhs_prefix = _generate_prefix("l") if not suffix_provided else ""
-    rhs_prefix = _generate_prefix("r") if not suffix_provided else ""
+    lhs_prefix = (
+        _generate_deterministic_prefix("l", all_names) if not suffix_provided else ""
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/dataframe.py:305]: why this is necessary?

File: src/snowflake/snowpark/dataframe.py [hunk 17]
```diff
@@ -610,21 +685,44 @@ def collect(
...
+                build_expr_from_dict_str_str(expr.statement_params, statement_params)
+            expr.block = block
+            expr.case_sensitive = case_sensitive
+            expr.log_on_exception = log_on_exception
+            expr.no_wait = False
+
+            self._session._ast_batch.eval(repr)
+
+            # Flush the AST and encode it as part of the query.
+            _, kwargs["_dataframe_ast"] = self._session._ast_batch.flush()
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/dataframe.py:706]: Can we create a constant for `"_dataframe_ast"`?

File: tests/.gitignore [hunk 18]
```diff
@@ -1 +1,2 @@
 parameters*.py
+cassettes/
```
Comment (sfc-gh-azhan) [inline tests/.gitignore:2]: curious what code generate this?

File: tests/integ/conftest.py [hunk 19]
```diff
@@ -21,6 +24,32 @@
...
+    "x-amz-server-side-encryption-customer-key-md5",
+    "x-amz-server-side-encryption-customer-key",
+    "x-amz-server-side-encryption-customer-algorithm",
+    "x-amz-id-2",
+    "x-amz-request-id",
+    "x-amz-version-id",
+]
+
+
+def _process_response_recording(response):
```
Comment (sfc-gh-azhan) [inline tests/integ/conftest.py:45]: where do we use _process_response_recording?

File: tests/integ/modin/test_apply_snowpark_python_functions.py [hunk 20]
```diff
@@ -33,29 +33,23 @@ def test_apply_sin():
 
 @sql_count_checker(query_count=4)
 def test_apply_log10():
-    from snowflake.snowpark.functions import log
+    from snowflake.snowpark.functions import _log10
```
Comment (sfc-gh-azhan) [inline tests/integ/modin/test_apply_snowpark_python_functions.py:36]: Is this a behavior change?

File: tests/integ/modin/tools/test_to_datetime.py [hunk 21]
```diff
@@ -597,7 +597,7 @@ def test_to_datetime_pydatetime(self):
     @pytest.mark.parametrize(
         "dt", [np.datetime64("2000-01-01"), np.datetime64("2000-01-02")]
     )
-    @sql_count_checker(query_count=1)
+    @sql_count_checker(query_count=1, join_count=0)
```
Comment (sfc-gh-azhan) [inline tests/integ/modin/tools/test_to_datetime.py:600]: shouldn't change this file

File: CHANGELOG.md (hunk 11 above)
Comment (sfc-gh-lspiegelberg) [reply CHANGELOG.md:22]: change to
`Added parameter `ast_enabled` to session for internal usage (default: False)`

File: src/snowflake/snowpark/query_history.py (hunk 12 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/query_history.py:96]: @sfc-gh-oplaton can we add a comment here?

File: src/snowflake/snowpark/modin/plugin/utils/numpy_to_pandas.py (hunk 13 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/modin/plugin/utils/numpy_to_pandas.py:216]: fix for broken code path in pandas. @sfc-gh-oplaton we can maybe add to umbrella ticket a follow up or file pandas bug.

File: src/snowflake/snowpark/column.py (hunk 14 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/column.py:257]: changes were reviewed by snowpark python team.

File: src/snowflake/snowpark/column.py (hunk 15 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/column.py:313]: offline: leave as-is.

File: src/snowflake/snowpark/dataframe.py (hunk 17 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/dataframe.py:706]: add `DATAFRAME_AST_PARAMETER="_dataframe_ast"`

File: tests/.gitignore (hunk 18 above)
Comment (sfc-gh-lspiegelberg) [reply tests/.gitignore:2]: remove this.

File: tests/integ/conftest.py (hunk 19 above)
Comment (sfc-gh-lspiegelberg) [reply tests/integ/conftest.py:45]: remove this. double-check this file.

File: tests/integ/modin/tools/test_to_datetime.py (hunk 21 above)
Comment (sfc-gh-lspiegelberg) [reply tests/integ/modin/tools/test_to_datetime.py:600]: revert this change.

**Comment by sfc-gh-lspiegelberg** [comment]
Add coverage to AST gates
```
      - name: Combine coverages
        run: python -m tox -e coverage --skip-missing-interpreters false
        shell: bash
        env:
          SNOWFLAKE_IS_PYTHON_RUNTIME_TEST: 1
      - uses: actions/upload-artifact@v4
        with:
          include-hidden-files: true
          name: coverage_${{ matrix.os }}-${{ matrix.python-version }}-${{ matrix.cloud-provider }}
          path: |
            .tox/.coverage
            .tox/coverage.xml
```

File: src/snowflake/snowpark/modin/plugin/utils/numpy_to_pandas.py (hunk 13 above)
Comment (sfc-gh-jkew) [reply src/snowflake/snowpark/modin/plugin/utils/numpy_to_pandas.py:216]: We will still encounter this form of snowpark function in the future. I'm fine with creating a new snowpark function that circumvents it for now; but the ast should handle this eventually.

File: src/snowflake/snowpark/_internal/analyzer/table_function.py [hunk 22]
```diff
@@ -88,7 +88,7 @@ def __init__(
         self.path = path
         self.outer = outer
         self.recursive = recursive
-        self.mode = mode
+        self.mode = mode.upper()
```
Comment (sfc-gh-jkew) [inline src/snowflake/snowpark/_internal/analyzer/table_function.py:91]: mode should be one of "OBJECT", "ARRAY", "BOTH" and typically the caller has been upper-ing this.

instead of changing this to mode.upper() you might change the signature to Literal["OBJECT", "ARRAY", "BOTH"]

File: src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py [hunk 23]
```diff
@@ -178,6 +178,7 @@ def _update_parents(
                 _update_parents(
                     node, should_replace_child=True, new_child=resolved_with_block
                 )
+                self._total_number_ctes += 1
```
Comment (sfc-gh-jkew) [inline src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py:181]: Hmm. Something is off here; this change is already in main afaict.

File: tests/.gitignore (hunk 18 above)
Comment (sfc-gh-lspiegelberg) [reply tests/.gitignore]: done.

File: tests/integ/modin/tools/test_to_datetime.py (hunk 21 above)
Comment (sfc-gh-lspiegelberg) [reply tests/integ/modin/tools/test_to_datetime.py]: done.

File: src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py (hunk 23 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py]: thanks for spotting, I restored this file to its state from main. I hope this doesn't break the tests.

File: tests/integ/conftest.py (hunk 19 above)
Comment (sfc-gh-lspiegelberg) [reply tests/integ/conftest.py]: reviewed file and removed dead code.

File: src/snowflake/snowpark/_internal/analyzer/table_function.py (hunk 22 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/_internal/analyzer/table_function.py:91]: this is a refactor we did approved by the snowpark team. I agree that this would be the correct way, yet sometimes literals like `object` or so are used. By adding the logic here to store a well-formed string, we were able to simplify logic in the corresponding functions. (Before it was something like `if self.mode.upper() == "OBJECT" ...`)

File: src/snowflake/snowpark/mock/_snowflake_data_type.py [hunk 24]
```diff
@@ -451,11 +523,33 @@ def __init__(self, *args, **kwargs) -> None:
...
         self._null_rows_idxs = []
 
-    def set_sf_type(self, value):
-        self.sf_type = value
+    @property
+    def sf_type(self) -> ColumnType:
+        if self._sf_type is not None:
+            return self._sf_type
+
+        # TODO: Else branch is taken when using UDTFs.
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/mock/_snowflake_data_type.py:531]: What is the TODO here?

**Review by sfc-gh-jrose — COMMENTED** [review]
Local testing section is looking mostly good to me.

File: tox.ini (hunk 2 above)
Comment (sfc-gh-lspiegelberg) [reply tox.ini:82]: Done, removed lines 84 and 84 because they're duplicated.

**Review by sfc-gh-jkew — APPROVED** [review]
I don't think this needs an intensive review - most of this will work itself out over the next day anyways.

File: src/snowflake/snowpark/mock/_snowflake_data_type.py (hunk 24 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/mock/_snowflake_data_type.py]: added to umbrella ticket SNOW-1826001

File: src/snowflake/snowpark/mock/_plan.py [hunk 25]
```diff
@@ -610,6 +899,9 @@ def execute_mock_plan(
 
         from_df = execute_mock_plan(from_, expr_to_alias)
 
+        if from_df is None:
+            return pandas.DataFrame()
```
[the remark this answers is not in the document]
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/mock/_plan.py]: changed to empty TableEmulator.

**Review by sfc-gh-azhan — APPROVED** [review]

**Review by sfc-gh-jrose — APPROVED** [review]

**Review by sfc-gh-aalam — CHANGES_REQUESTED** [review]

File: src/snowflake/snowpark/_internal/analyzer/table_merge_expression.py [hunk 26]
```diff
@@ -38,10 +39,10 @@ def individual_node_complexity(self) -> Dict[PlanNodeCategory, int]:
 
 class UpdateMergeExpression(MergeExpression):
     def __init__(
-        self, condition: Optional[Expression], assignments: Dict[Expression, Expression]
+        self, condition: Optional[Expression], assignments: Dict[str, ColumnOrLiteral]
```
[the remark this answers is not in the document]
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/_internal/analyzer/table_merge_expression.py:42]: yes, the API calling the constructor here has
` def update(self, assignments: Dict[str, ColumnOrLiteral]) -> "WhenMatchedClause"`.
Hence, this type hint. Dict[Expression, Expression] seems wrong.

File: src/snowflake/snowpark/_internal/proto/update-from-devvm.sh [hunk 27]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+# This is a little helper script to build the python protobuf file and copy it over to this directory.
+# If not pulling from the devvm, please provide the SSH Host as an argument.
+# Cloud workspace ssh is set up via ~/.local/share/sfcli/ssh_config (or simply provide your workspace ID)
+# For this script to succeed with the DevVM, you need to configure the SSH devvm host according to
+# https://snowflakecomputing.atlassian.net/wiki/spaces/EN/pages/3019015180/DevVM+Troubleshooting#Symptom:-Need-to-ssh-into-the-certified-DevVM
+# I.e., add the following lines to ~/.ssh/config
```
[the remark this answers is not in the document]
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/_internal/proto/update-from-devvm.sh:7]: sure, which information should be redacted in your opinion?

File: src/snowflake/snowpark/session.py [hunk 28]
```diff
@@ -745,6 +809,28 @@ def auto_clean_up_temp_table_enabled(self) -> bool:
         :meth:`DataFrame.cache_result` in the current session when the DataFrame is no longer referenced (i.e., gets garbage collected).
         The default value is ``False``.
 
+        Example::
+
+            >>> import gc
+            >>>
+            >>> def f(session: Session) -> str:
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/session.py:816]: this is also unrealted to your change it seems

File: src/snowflake/snowpark/session.py [hunk 29]
```diff
@@ -2738,22 +2946,26 @@ def write_pandas(
...
-                    parallel=parallel,
-                    quote_identifiers=quote_identifiers,
-                    auto_create_table=auto_create_table,
-                    overwrite=overwrite,
-                    table_type=table_type,
-                    **kwargs,
-                )
+                if isinstance(self._conn, MockServerConnection):
+                    # TODO: Implement here write_pandas correctly.
+                    success, ci_output = True, []
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/session.py:2951]: we already raise a not implemented error earlier. Does adding an assert here make more sense?

File: src/snowflake/snowpark/table.py [hunk 30]
```diff
@@ -222,11 +236,11 @@ def insert(
                 "insert", "WhenNotMatchedClause"
             )
         if isinstance(assignments, dict):
-            keys = [Column(k)._expression for k in assignments.keys()]
-            values = [Column._to_expr(v) for v in assignments.values()]
+            keys = list(assignments.keys())
+            values = list(assignments.values())
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/table.py:240]: I think this answers my earlier question

File: src/snowflake/snowpark/udf.py [hunk 31]
```diff
@@ -602,7 +628,7 @@ def register(
         - :meth:`register_from_file`
         """
         with open_telemetry_udf_context_manager(self.register, func=func, name=name):
-            if not callable(func):
+            if not callable(func) and kwargs.get("_registered_object_name") is None:
                 raise TypeError(
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/udf.py:632]: is this AST only change?

File: src/snowflake/snowpark/_internal/proto/update-from-devvm.sh (hunk 27 above)
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/proto/update-from-devvm.sh:7]: ssh and sfcli related information

File: src/snowflake/snowpark/udf.py (hunk 31 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/udf.py:632]: @sfc-gh-evandenberg has some more context on this. Basically, this a non-public parameter which helps to replay ASTs.

File: src/snowflake/snowpark/session.py (hunk 28 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/session.py:816]: https://snowflakecomputing.atlassian.net/browse/SNOW-1842532

File: src/snowflake/snowpark/_internal/proto/update-from-devvm.sh (hunk 27 above)
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/proto/update-from-devvm.sh:7]: https://snowflakecomputing.atlassian.net/browse/SNOW-1842532

File: src/snowflake/snowpark/session.py (hunk 29 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/session.py:2951]: https://snowflakecomputing.atlassian.net/browse/SNOW-1842541
